### ISSUE TITLE
Socket unsub is broken for Trade Streams

### ISSUE THREAD

**Issue by HypsyNZ** [JKorf/Binance.Net#1179]
https://github.com/JKorf/Binance.Net/blob/81293c84823903f5e824ad1ef96b97ce52989869/Binance.Net/Clients/SpotApi/BinanceSocketClientSpotStreams.cs#L547

`NextId()` is wrong here, it should be the unique `Id` from the `Request` like below?

```cs
 var unsub = new BinanceSocketRequest
            {
                Method = "UNSUBSCRIBE",
                Params = Request.Params,
                Id = Request.Id,
            };
```

Currently if you `unsubscribe` from some things you will keep receiving data (Trades, Orderbook)

**Comment by Hulkstance** [comment]
No, it's not wrong. You are creating a new request for the unsubscribe. https://github.com/binance/binance-spot-api-docs/blob/master/web-socket-streams.md#unsubscribe-to-a-stream

It works fine. You can literally see the "Connection closed" message. You did something wrong.

```cs
var subscriptionResult = await _socketClient.SpotStreams.SubscribeToOrderBookUpdatesAsync("BTCUSDT",
    null, data =>
    {
        _logger.LogInformation("Timestamp: {Timestamp}", data.Timestamp);
    });

if (!subscriptionResult.Success)
{
    Console.WriteLine("Failed to connect: " + subscriptionResult.Error);
    return;
}

subscriptionResult.Data.ConnectionClosed += () =>
{
    Console.WriteLine("Connection closed");
};
subscriptionResult.Data.ConnectionLost += () =>
{
    Console.WriteLine("Connection lost");
};
subscriptionResult.Data.ConnectionRestored += (time) =>
{
    Console.WriteLine("Connection restored");
};

await Task.Delay(5000);

await _socketClient.UnsubscribeAsync(subscriptionResult.Data.Id);

// I would rather do:
// await subscriptionResult.Data.CloseAsync();
```

**Comment by HypsyNZ** [comment]
Sorry about this @Hulkstance, The issue I was having was from changes I made to the library.

Thank you for testing.

**Comment by HypsyNZ** [comment]
https://github.com/binance/binance-spot-api-docs/blob/master/web-socket-streams.md#unsubscribe-to-a-stream

The `ID` should be the same `ID` from the `Subscribe` Request

Subscirbe: 
```cs
{
  "method": "SUBSCRIBE",
  "params": [
    "btcusdt@depth"
  ],
  "id": 7777
}
```

Unsubscribe:
```cs
{
  "method": "UNSUBSCRIBE",
  "params": [
    "btcusdt@depth"
  ],
  "id": 7777
}
```

I keep receiving data from the websocket if I don't change this even though it says it was "successful"

**Comment by JKorf** [comment]
The id is just a message id to link the message sent to a response from the server, it shouldn't be the same id as the subscribe message. I can however confirm that the trade stream doesn't stop data, even though the server responds with a success message to the unsubscribe request.

I've opened an issue for this on the Binance forum, seems to be an issue on their end:
https://dev.binance.vision/t/unsubscribing-trades-stream-doesnt-stop-data/14322

**Comment by HypsyNZ** [comment]
> The id is just a message id to link the message sent to a response from the server, it shouldn't be the same id as the subscribe message. I can however confirm that the trade stream doesn't stop data, even though the server responds with a success message to the unsubscribe request.
> 
> I've opened an issue for this on the Binance forum, seems to be an issue on their end: https://dev.binance.vision/t/unsubscribing-trades-stream-doesnt-stop-data/14322

If the `ID` matches in both requests it will work, I've already done it.

**Comment by JKorf** [comment]
Then why does the unsubscribing work for other streams? And why does the server return a success message?
I only see this line about the Id:
`The id used in the JSON payloads is an unsigned INT used as an identifier to uniquely identify the messages going back and forth.`
I don't see anywhere that the unsubscribe message should have the same id as the subscribe message. But it's weird that it would work

**Comment by HypsyNZ** [comment]
> Then why does the unsubscribing work for other streams? And why does the server return a success message? I only see this line about the Id: `The id used in the JSON payloads is an unsigned INT used as an identifier to uniquely identify the messages going back and forth.` I don't see anywhere that the unsubscribe message should have the same id as the subscribe message. But it's weird that it would work

I don't know either but I changed it so they match like in my example and now it seems to work as intended. 

_shrug_

**Comment by Hulkstance** [comment]
How did you reproduce it? It works fine for me on 8.3.0.

**Comment by HypsyNZ** [comment]
> Could you provide a minimal reproducible example?

I already fixed the bug here is working version.

I use a random ID for the new request and then provide the ID when unsubscribing.

[Binance-Trader SocketClientHost](
https://github.com/HypsyNZ/Binance-Trader/blob/master/BinanceTrader/BVVM/BT/Binance-API/ClientHosts/SocketClientHost.cs#L95)

> It works fine for me on 8.3.0.

are you using `Trade Streams` I can clearly see its wrong in the code base.

**Comment by JKorf** [comment]
@Hulkstance 
I can reproduce it like this:
```C#
var x = await binanceSocketClient.SpotStreams.SubscribeToTradeUpdatesAsync("BTCUSDT", data =>
{
    Console.WriteLine("data");
});

var y = await binanceSocketClient.SpotStreams.SubscribeToBookTickerUpdatesAsync("BTCUSDT", data => { });

Console.ReadLine();

await x.Data.CloseAsync();
Console.ReadLine();
````
Note that I start 2 streams on the same socket. If you only start one stream the socket will disconnect after unsubscribing, which means that the stream automatically stops. By having another subscription on the same socket the socket will remain open after unsubscribing the trade stream and you can see data will continue to come in from the trade stream.
If you put LogLevel on Trace in the client options you can see what's going on.

@HypsyNZ lets see what the response is on the forum post

**Comment by Hulkstance** [comment]
@JKorf, works fine for me.

![image](https://user-images.githubusercontent.com/33224919/210370571-94c2e916-e43b-4bd5-bf7f-58b3053f0d2d.png)

You can see that the connection remains opened in Process Hacker.

**Comment by HypsyNZ** [comment]
> @JKorf, works fine for me. You can see that the connection remains opened in Process Hacker.

So you were able to reproduce it?

**Comment by Hulkstance** [comment]
> > @JKorf, works fine for me. You can see that the connection remains opened in Process Hacker.
> 
> So you were able to reproduce it?

My bad. I was able to reproduce it:

```cs
using Binance.Net.Clients;
using CryptoExchange.Net.Logging;
using Microsoft.Extensions.Logging;

var binanceSocketClient = new BinanceSocketClient
{
    ClientOptions =
    {
        LogLevel = LogLevel.Trace,
        LogWriters = new List<ILogger>
        {
            new ConsoleLogger()
        }
    }
};

var x = await binanceSocketClient.SpotStreams.SubscribeToTradeUpdatesAsync("BTCUSDT", data =>
{
    Console.WriteLine($"Timestamp: {data.Timestamp}");
});

var y = await binanceSocketClient.SpotStreams.SubscribeToBookTickerUpdatesAsync("BTCUSDT", data => { });

Console.ReadLine();

await x.Data.CloseAsync();
Console.WriteLine("Stopped");

Console.ReadLine();
```

**Comment by HypsyNZ** [comment]
I've modified quite a bit so I'm not sure how easy it will be to "just test" changing it to the `request ID` but It seems like it worked.

**Comment by HypsyNZ** [comment]
Just noticed this `Trade Stream` is sending incomplete data?

can anyone else confirm this?

**Comment by HypsyNZ** [comment]
I'm still receiving the bad data I'm just ignoring the message handler

**Comment by HypsyNZ** [comment]
Just a stab in the dark but could this be because of the keep alive thing?
